README.md
<h1>Welcome To LEAF Project</h1>
<b>L</b>ightweight <b>E</b>asy <b>A</b>pplication <b>F</b>ramework

<h5>Leaf is an open source application development framework for PHP.</h5>

<p>
  Version : 1.2.1 <br>
  Link: <a target="_blank" href="http://amitpatil.net/leaf">http://amitpatil.net/leaf</a><br><br>
</p>

<p>
  Leaf is a small scale PHP Application Development Framework, built on basic principle 
  of REST. The initial idea behind leaf project is to build PHP Web Services using very few lines of code. Leaf tries to make MVC Architecture easy to understand and implement.
</p>

<p>
  For details and documentation <a target="_blank" href="http://amitpatil.net/leaf/docs">click here.</a>
</p>



changes.txt
Dec 10, 2015:
Only GET and POST requests handled by REQUEST parameter:
	LEAF only focuses on two HTTP requests i.e. GET and POST.
	All other requests are filtered.
Dec 1, 2015:
Multi Key-Value pair authentication:	
	Unlike previous version in this version REQUEST can be authenticated using different different key-value pair.
	These key-value pairs can be plainly stored in auth.php file or can be fetched from database.
Nov 29, 2015:
Multiple database selection:
	User can select multiple database from dbConfig.php by just passing different db_id.


application/controllers/cp_test.php
<?php

/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of cp_test
 *
 * Page is created for understanding only.
 * 
 * @author Amit
 */
class test {

    function index_get() {
        include_once VIEW_PATH . '_header.php';
        include_once VIEW_PATH . 'welcome.php';
        include_once VIEW_PATH . '_footer.php';
    }

    function show_rows_get() {
        require_once MODEL_PATH . 'testDBAccessModel.php';
        $db_access = new TestDBAccessModel();

        /*
         * http://localhost/LEAF/test/show_rows?iformat=json
         * 
         */
        
        return $db_access->select_all_rows();
        
        //--------------------OR------------------------------
        
        
        /*
         * http://localhost/LEAF/test/show_rows?iformat=text
         * 
         */
        
        //return $db_access->select_all_rows()[0]['column'];
        
        //--------------------OR------------------------------
   
        /*
         * http://localhost/LEAF/test/show_rows
         * OR
         * http://localhost/LEAF/test/show_rows?iformat=html         * 
         * 
         */
        
        //require_once VIEW_PATH . 'testRowView.php';
        //$row_view = new TestRowView();

        //include_once VIEW_PATH . '_header.php';
        //$row_view->_body($db_access->select_all_rows());
        //include_once VIEW_PATH . '_footer.php';
    }

}


system/auth.php
<?php

/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of auth.php
 *
 * HTTP request authentication page
 * 
 * @author Amit
 */

if (API_KEY === 'yes') {
    // Check HTTP request and collect key and value
    if (REQUEST === 'POST') {
        if (!is_null(filter_input(INPUT_POST, 'ikey')) && ctype_alnum(filter_input(INPUT_POST, 'ikey')) && !is_null(filter_input(INPUT_POST, 'ipass')) && ctype_alnum(filter_input(INPUT_POST, 'ipass'))) {
            $key = filter_input(INPUT_POST, 'ikey', FILTER_SANITIZE_STRING);
            $value = filter_input(INPUT_POST, 'ipass', FILTER_SANITIZE_STRING);
        } else {
            header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
            echo 'You are not authorised to access this page.';
            exit(1); // EXIT_ERROR 
        }
    } elseif (REQUEST === 'GET') {
        if (!is_null(filter_input(INPUT_GET, 'ikey')) && ctype_alnum(filter_input(INPUT_GET, 'ikey')) && !is_null(filter_input(INPUT_GET, 'ipass')) && ctype_alnum(filter_input(INPUT_GET, 'ipass'))) {
            $key = filter_input(INPUT_GET, 'ikey', FILTER_SANITIZE_STRING);
            $value = filter_input(INPUT_GET, 'ipass', FILTER_SANITIZE_STRING);
        } else {
            header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
            echo 'You are not authorised to access this page.';
            exit(1); // EXIT_ERROR 
        }
    } else {
        header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
        echo 'You are not authorised to access this page.';
        exit(1); // EXIT_ERROR 
    }
    //Depending on authentication method validate key-value pair
    if (AUTH_TYPE === 'plain') {
        $authorizer = array(
            'user1' => '65h4g9874h65464h894hg65487h64',
            'user2' => 'gfd46165fdh897yt6541616df7rer',
            'user3' => 'gg8971646uo561dfh546d78te564r',
            'user4' => 'uku55fh156456j3213gd1651af654'
        );
        if (strcmp($authorizer[$key], $value) !== 0) {
            header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
            echo 'You are not authorised to access this page.';
            exit(1); // EXIT_ERROR 
        }
    } elseif (AUTH_TYPE === 'db') {
        require_once SYS_PATH . 'dbConfig.php';
        $dbc = new DBC(0); // 0 defines database id in dbConfig.php
        $db = $dbc->getConnection();
        $db->exec("SET CHARACTER SET utf8");
        try {
            $sql = "SELECT id FROM `users` WHERE `key`= ? AND `pass`= ?";
            $stmt = $db->prepare($sql);
            $stmt->execute(array($key, $value));
            $stmt->setFetchMode(PDO::FETCH_ASSOC);
            $rows = $stmt->fetchAll();
            if (empty($rows)) {
                header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
                echo 'You are not authorised to access this page.';
                exit(1); // EXIT_ERROR 
            }
            $dbc->closeConnection();
        } catch (PDOException $e) {
            header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
            echo 'You are not authorised to access this page.';
            exit(1); // EXIT_ERROR 
        }
    } else {
        header('HTTP/1.1 401 Unauthorized.', TRUE, 401);
        echo 'You are not authorised to access this page.';
        exit(1); // EXIT_ERROR  
    }
}


system/config.php
<?php
/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of config.php
 *
 * Main configuration page
 * 
 * @author Amit
 */

/*
 * --------------------------------------------------------------------
 * URL
 * --------------------------------------------------------------------
 * 
 */

define('WEB_URL', 'http://localhost/LEAF/');


/*
 * ---------------------------------------------------------------
 * API KEY
 * ---------------------------------------------------------------
 * 
 * API Key is case-sensetive. 
 * 
 * Disable API Key checking by keeping API_KEY blank.
 * 
 * If enabled, check .htaccess file for some changes
 * 
 */

define('API_KEY', 'no');


/*
 * ---------------------------------------------------------------
 * AUTHENTICATION TYPE
 * ---------------------------------------------------------------
 * 
 * Authentication Type
 * 
 * plain : Use array of key => value from auth.php file
 * db    : Use table of key and value pair in database
 * 
 */

define('AUTH_TYPE', 'db');

/*
 * --------------------------------------------------------------------
 * CONTROLLER FILE PREFIX
 * --------------------------------------------------------------------
 * 
 * Prefix in controller file is used for security reason only, keep blank if not required.
 * If provided, add it to each controllers name.
 * 
 */

define('CONTROLLER_PREFIX', 'cp_');

/*
 * --------------------------------------------------------------------
 * CLASS NAME CASE
 * --------------------------------------------------------------------
 * 
 * Prefix in controller file is used for security reason only, keep blank if not required.
 * If provided, add it to each controllers name.
 * 
 */

define('CHANGE_CLASS_CASE', TRUE);


/*
 * --------------------------------------------------------------------
 * DEFAULT TIME ZONE
 * --------------------------------------------------------------------
 */
date_default_timezone_set("Asia/Kolkata");


/*
 * --------------------------------------------------------------------
 * DEFAULT RESPONSE FORMAT
 * --------------------------------------------------------------------
 */
define('DEFAULT_RESPONSE', 'html');

/*
 * --------------------------------------------------------------------
 * ALLOWED RESPONSES
 * --------------------------------------------------------------------
 */
$allowed_responses = array('html', 'text', 'json');


system/dbConfig.php
<?php

/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of dbConfig.php
 *
 * Database selection and connection page
 * 
 * @author Amit
 */

/*
 * --------------------------------------------------------------------
 * DATABASE DRIVERS
 * --------------------------------------------------------------------
 * 
 * Defult Database settings to be used.
 * 
 * Include any number of database connection and pass $db_id to use it.
 * 
 */

class DBC {

    var $db_id = 0;
    var $config = array(
        0 => array(
            'driver' => 'PDO',
            'host' => '',
            'user' => '',
            'password' => '',
            'database' => ''),
        1 => array(
            'driver' => 'OOP',
            'host' => '',
            'user' => '',
            'password' => '',
            'database' => ''),
        2 => array(
            'driver' => 'FUNC',
            'host' => '',
            'user' => '',
            'password' => '',
            'database' => '')
    );
    var $db = null;

    function __construct($db_id = 0) {
        $this->db_id = $db_id;
        switch ($this->config[$db_id]['driver']) {
            case 'PDO':
                $this->db = new PDO("mysql:host=" . $this->config[$db_id]['host'] . ";dbname=" . $this->config[$db_id]['database'], $this->config[$db_id]['user'], $this->config[$db_id]['password']);
                break;
            case 'OOP':
                $this->db = new mysqli($this->config[$db_id]['host'], $this->config[$db_id]['user'], $this->config[$db_id]['password'], $this->config[$db_id]['database']);
                break;
            case 'FUNC':
                $this->db = mysqli_connect($this->config[$db_id]['host'], $this->config[$db_id]['user'], $this->config[$db_id]['password'], $this->config[$db_id]['database']);
                break;
        }
    }

    function getConnection() {
        return $this->db;
    }

    function closeConnection() {
        if (!empty($this->db)) {
            switch ($this->config[$this->db_id]['driver']) {
                case 'PDO':
                    $this->db = NULL;
                    break;
                case 'OOP':
                    $this->db->close();
                    break;
                case 'FUNC':
                    mysqli_close($this->db);
                    break;
            }
        }
    }

}


system/httpResponse.php
<?php
/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of httpResponse
 *
 * Build HTTP response
 * 
 * @author Amit
 */

/*
 * ---------------------------------------------------------------
 * FUNCTION RESPONSE
 * ---------------------------------------------------------------
 * 
 */

function _response($code, $format, $status, $obj = null, $method = null, $paraArray = null) {

    // Define HTTP responses
    $http_response_code = array(
        200 => 'OK',
        400 => 'Bad Request',
        401 => 'Unauthorized',
        403 => 'Forbidden',
        404 => 'Not Found'
    );

    // Set HTTP Response
    header('HTTP/1.1 ' . $status . ' ' . $http_response_code[$status]);
    header("Cache-Control: no-cache, must-revalidate");

    // Set HTTP Response Format
    switch ($format) {
        case 'text':
            header('Content-Type: text/plain; charset=utf-8');
            break;
        case 'json':
            header('Content-Type: application/json; charset=utf-8');
            break;
        case 'html':
        default:
            header('Content-Type: text/html; charset=utf-8');
            break;
    }

    // Check Success code
    if ($code === 1) {
        if (method_exists($obj, $method)) {
            $result = call_user_func_array(array($obj, $method), $paraArray);
        } else {
            $status = 400;
            // Replacement header
            header('HTTP/1.1 ' . $status . ' ' . $http_response_code[$status]);
            $result = "Requested URL Not Found.";
        }
    } else {
        // Custom Status Messages
        switch ($status) {
            case 400:
                $result = "Requested URL Not Found.";
                break;
            case 401:
                $result = "You Are Not Authorized To Access This Page.";
                break;
            case 403:
                $result = "You Are Not Authorized To Access This Page.";
                break;
            case 404:
                $result = "Requested Page Not Found.";
                break;
            default:
                $result = "Failed To Process Your Request.";
                break;
        }
    }
    switch ($format) {
        case 'text':
            echo $result;
            break;
        case 'json':
            echo json_encode($result);
            break;
        case 'html':
        default:
            echo $result;
            break;
    }
    exit;
}


application/controllers/cp_welcome.php
<?php

/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of cp_welcome
 * 
 * Home page for Leaf Project
 *
 * @author Amit
 */
class welcome {

    function index_get() {
        include_once VIEW_PATH . '_header.php';
        include_once VIEW_PATH . 'welcome.php';
        include_once VIEW_PATH . '_footer.php';
    }

}


application/models/testDBAccessModel.php
<?php

/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of testDBAccessModel
 *
 * Page is created for understanding only.
 * 
 * @author Amit
 */
require_once SYS_PATH . 'dbConfig.php';

class TestDBAccessModel {

    var $db = null;

    function __construct() {
        $dbc = new DBC(0);
        $this->db = $dbc->getConnection();
        $this->db->exec("SET CHARACTER SET utf8");
    }
    
    function select_all_rows() {
        try {
            $sql = "SELECT * FROM `table` LIMIT 50";
            $stmt = $this->db->prepare($sql);
            $stmt->execute();
            $stmt->setFetchMode(PDO::FETCH_ASSOC);
            $rows = $stmt->fetchAll();
            return $rows;
        } catch (PDOException $e) {
            return FALSE;
        }
    }

}


index.php
<?php

/*
 * LEAF (Lightweight Easy Application Framework)
 *
 * An open source application development framework for PHP
 *
 * This content is released under the MIT License (MIT)
 *
 * Copyright (c) 2015 - 2016 AmitPatil Codes.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *  @package   Leaf
 *  @copyright Copyright (c) 2015 - 2016, AmitPatil Codes
 *  @link  http://amitpatil.net/leaf
 *  @since	Version 1.2.1
 */

/**
 * Description of index.php
 *
 * Main page accepting user request
 * 
 * @author Amit
 */

/*
 * ---------------------------------------------------------------
 * APPLICATION ENVIRONMENT
 * ---------------------------------------------------------------
 *
 *     development
 *     testing
 *     production
 *     maintenance
 *
 */

define('ENVIRONMENT', isset($_SESSION['APP_ENV']) ? $_SESSION['APP_ENV'] : 'development');

/*
 * ---------------------------------------------------------------
 * ERROR REPORTING
 * ---------------------------------------------------------------
 */

switch (ENVIRONMENT) {
    case 'development':
        error_reporting(-1);
        ini_set('display_errors', 1);
        break;
    case 'testing':
    case 'production':
        ini_set('display_errors', 0);
        if (version_compare(PHP_VERSION, '5.3', '>=')) {
            error_reporting(E_ALL & ~E_NOTICE & ~E_DEPRECATED & ~E_STRICT & ~E_USER_NOTICE & ~E_USER_DEPRECATED);
        } else {
            error_reporting(E_ALL & ~E_NOTICE & ~E_STRICT & ~E_USER_NOTICE);
        }
        break;
    case 'maintenance':
        header('HTTP/1.1 503 Service Unavailable.', TRUE, 503);
        echo 'Sorry for the inconvenience but we&rsquo;re performing some maintenance at the moment. We&rsquo;ll be back online shortly!';
        exit(1); // EXIT_ERROR
        break;
    default:
        header('HTTP/1.1 503 Service Unavailable.', TRUE, 503);
        echo 'The application environment is not set correctly.';
        exit(1); // EXIT_ERROR
}

/*
 * ---------------------------------------------------------------
 * FOLDERS AND PATH
 * ---------------------------------------------------------------
 */
define('APP_PATH_SEPARATOR', '/');

$system_dir = 'system';
$application_dir = 'application';
$controller_dir = 'controllers';
$models_dir = 'models';
$view_dir = 'views';

define('BASE_PATH', __DIR__ . APP_PATH_SEPARATOR);
define('SYS_PATH', BASE_PATH . $system_dir . APP_PATH_SEPARATOR);
define('APP_PATH', BASE_PATH . $application_dir . APP_PATH_SEPARATOR);
define('CONTROLLER_PATH', APP_PATH . $controller_dir . APP_PATH_SEPARATOR);
define('MODEL_PATH', APP_PATH . $models_dir . APP_PATH_SEPARATOR);
define('VIEW_PATH', APP_PATH . $view_dir . APP_PATH_SEPARATOR);

if (!is_dir(SYS_PATH) || !is_dir(APP_PATH) || !is_dir(CONTROLLER_PATH) || !is_dir(MODEL_PATH) || !is_dir(VIEW_PATH)) {

    header('HTTP/1.1 404 Not Found.', TRUE, 404);
    echo 'The Application can not find specified directory or path.';
    exit(3); // EXIT_CONFIG
}

/*
 * --------------------------------------------------------------------
 * CHECK FOR HTTP REQUEST
 * --------------------------------------------------------------------
 * 
 */
define('REQUEST', filter_input(INPUT_SERVER, 'REQUEST_METHOD'));

/*
 * --------------------------------------------------------------------
 * LOAD CORE FILES
 * --------------------------------------------------------------------
 */
require_once SYS_PATH . 'config.php';
require_once SYS_PATH . 'auth.php';
require_once SYS_PATH . 'httpResponse.php';

// Set default format
$format = DEFAULT_RESPONSE;

// Function call
if (REQUEST === 'POST') {
    $format = !is_null(filter_input(INPUT_POST, 'iformat')) && in_array(filter_input(INPUT_POST, 'iformat'), $allowed_responses) ? filter_input(INPUT_POST, 'iformat') : DEFAULT_RESPONSE;
    if (!is_null(filter_input(INPUT_POST, 'ipage')) && preg_match('/[a-zA-Z0-9_]/', filter_input(INPUT_POST, 'ipage'))) {
        // Can replace strtolower() function with strtoupper().
        $class = CHANGE_CLASS_CASE ? strtolower(filter_input(INPUT_POST, 'ipage')) : filter_input(INPUT_POST, 'ipage');
        // Php (controller) file to be included
        $page = CONTROLLER_PREFIX . $class . '.php';
        if (file_exists(CONTROLLER_PATH . $page)) {
            include_once CONTROLLER_PATH . $page;
            if (!is_null(filter_input(INPUT_POST, 'imethod')) && preg_match('/[a-zA-Z0-9_]/', filter_input(INPUT_POST, 'imethod'))) {
                _response(1, $format, 200, new $class(), filter_input(INPUT_POST, 'imethod') . '_post', requestArrayFilter($_POST));
            } else {
                _response(0, $format, 400);
            }
        } else {
            _response(0, $format, 404);
        }
    } else {
        _response(0, $format, 400);
    }
} else {
    $format = !is_null(filter_input(INPUT_GET, 'iformat')) && in_array(filter_input(INPUT_GET, 'iformat'), $allowed_responses) ? filter_input(INPUT_GET, 'iformat') : DEFAULT_RESPONSE;
    if (!is_null(filter_input(INPUT_GET, 'ipage')) && preg_match('/[a-zA-Z0-9_]/', filter_input(INPUT_GET, 'ipage'))) {
        // Can replace strtolower() function with strtoupper().
        $class = CHANGE_CLASS_CASE ? strtolower(filter_input(INPUT_GET, 'ipage')) : filter_input(INPUT_GET, 'ipage');
        // Php (controller) file to be included
        $page = CONTROLLER_PREFIX . $class . '.php';
        if (file_exists(CONTROLLER_PATH . $page)) {
            include_once CONTROLLER_PATH . $page;
            if (!is_null(filter_input(INPUT_GET, 'imethod')) && preg_match('/[a-zA-Z0-9_]/', filter_input(INPUT_GET, 'imethod'))) {
                _response(1, $format, 200, new $class(), filter_input(INPUT_GET, 'imethod') . '_get', requestArrayFilter($_GET));
            } else {
                _response(0, $format, 400);
            }
        } else {
            _response(0, $format, 404);
        }
    } else {
        _response(0, $format, 400);
    }
}

// Filter reserve keywords from request string
function requestArrayFilter($requestArray) {
    $paraArray = array();
    foreach ($requestArray as $key => $value) {
        if (!in_array($key, array('ipage', 'imethod', 'ikey', 'ipass', 'iformat'))) {
            $paraArray[] = $value;
        }
    }
    return $paraArray;
}

// Function to debug
function logger($msg) {
    $file = fopen(BASE_PATH . "log.txt", "a");
    fwrite($file, date('D, d M Y H:i:s') . ": " . $msg . "\n");
    fclose($file);
}
